turbopack/crates/turbopack-core/src/resolve/mod.rs

```diff
@@ -3335,3 +3348,306 @@ impl Display for ModulePart {
+                ("./client", "client#component.js"),
+            ],
+        })
+        .await;
+    }
+
+    #[tokio::test(flavor = "multi_thread", worker_threads = 2)]
+    async fn test_fragment_with_ts_priority() {
+        // Fragment handling with extension priority
+        resolve_relative_request_test(ResolveRelativeRequestTestParams {
+            files: vec!["page#section.js", "page#section.ts"],
+            pattern: rcstr!("./page#section").into(),
+            enable_typescript_with_output_extension: true,
+            fully_specified: false,
+            expected: vec![("./page", "page#section.ts")],
+        })
+        .await;
+    }
+
+    // Dynamic pattern tests
+    #[tokio::test(flavor = "multi_thread", worker_threads = 2)]
+    async fn test_dynamic_pattern_with_js_extension() {
+        // Pattern: ./src/*.js should generate multiple keys with .ts priority
+        // When both foo.js and foo.ts exist, dynamic patterns need both keys for runtime resolution
+        // Results are sorted alphabetically by key
+        resolve_relative_request_test(ResolveRelativeRequestTestParams {
+            files: vec!["src/foo.js", "src/foo.ts", "src/bar.js"],
+            pattern: Pattern::Concatenation(vec![
+                Pattern::Constant(rcstr!("./src/")),
+                Pattern::Dynamic,
+                Pattern::Constant(rcstr!(".js")),
+            ]),
+            enable_typescript_with_output_extension: true,
+            fully_specified: false,
+            expected: vec![
+                // WRONG: request key doesn't match pattern
+                ("./src/foo.ts", "src/foo.ts"),
+                ("./src/bar.js", "src/bar.js"),
+                // WRONG: source file is redundant with extension rewriting
+                ("./src/foo.js", "src/foo.js"),
+            ],
+        })
+        .await;
+    }
+
+    #[tokio::test(flavor = "multi_thread", worker_threads = 2)]
+    async fn test_dynamic_pattern_without_extension() {
+        // Pattern: ./src/* (no extension) with TypeScript priority
+        // Dynamic patterns generate keys for all matched files, including extension alternatives
+        // Results are sorted alphabetically by key
+        resolve_relative_request_test(ResolveRelativeRequestTestParams {
+            files: vec!["src/foo.js", "src/foo.ts", "src/bar.js"],
+            pattern: Pattern::Concatenation(vec![
+                Pattern::Constant(rcstr!("./src/")),
+                Pattern::Dynamic,
+            ]),
+            enable_typescript_with_output_extension: true,
+            fully_specified: false,
+            expected: vec![
+                ("./src/bar", "src/bar.js"),
+                ("./src/bar.js", "src/bar.js"),
+                // WRONG: all three should point at the .ts file
+                ("./src/foo", "src/foo.js"),
+                ("./src/foo.js", "src/foo.js"),
+                // WRONG: the request key should be .js
```

Comment:
this isn't actually wrong, we might expect the user to construct a `.ts` request string at runtime... maybe